File: src/DiagToolsValidator/DiagToolsValidator.Core/Configuration/LTTngTest/LTTngTestRunConfiguration.fs
namespace DiagToolsValidator.Core.Configuration

open System
open System.IO
open System.Collections.Generic
open System.Collections

open YamlDotNet.Serialization

open DiagToolsValidator.Core.Functionality

module LTTngTestRunConfiguration =
    type SystemInformation() =
        member val EnvironmentVariables = new Dictionary<string, string>() with get, set

    type DotNetSetting() =
        member val SDKVersion: string = null with get, set
        member val DotNetRoot: string = null with get, set

    type TargetAppSetting() =
        member val BuildConfig: string = null with get, set
        member val GCPerfsim: DotNetApp.DotNetApp = new DotNetApp.DotNetApp() with get, set
        
    type LTTngTestBaseConfiguration() =
        member val TestBed: string = null with get, set
        member val SDKVersionList: string list = List.empty with get, set
        member val TargetApp: TargetAppSetting = new TargetAppSetting() with get, set
        member val SystemInfo: SystemInformation = new SystemInformation() with get, set
   
    type LTTngTestRunConfiguration() =
        member val TestBed: string = null with get, set
        member val TestResultFolder: string = null with get, set
        member val PerfCollectPath: string = null with get, set
        member val DotNet: DotNetSetting = new DotNetSetting() with get, set
        member val TargetApp: TargetAppSetting = new TargetAppSetting() with get, set
        member val SystemInfo: SystemInformation = new SystemInformation() with get, set

    type LTTngTestConfiguration() =
        member val TestBed: string = null with get, set
        member val RunConfigurationList: LTTngTestRunConfiguration list = List.empty with get, set

    [<AbstractClass;Sealed>]
    type LTTngTestConfigurationGenerator private() =
        static let ParseConfigFile(path: string) = 
            let _deserializer = (new DeserializerBuilder()).IgnoreUnmatchedProperties().Build()
            let serializedConfiguration = File.ReadAllText(path)

            try
                let configuration = _deserializer.Deserialize<LTTngTestBaseConfiguration>(serializedConfiguration);
                let parseExn = new exn("Fail to parse config file")
                if List.isEmpty configuration.SDKVersionList
                then
                    parseExn.Data.Add(nameof(LTTngTestConfigurationGenerator), "Please specify .NET SDK version")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.TestBed
                then
                    parseExn.Data.Add(nameof(LTTngTestConfigurationGenerator), "Please specify testbed")
                    raise(parseExn)

                else 
                    configuration
            with ex -> 
                reraise()
         
        static member GenerateRunConfiguration(path: string) = 
            let baseConfig = ParseConfigFile path
            let aggregatedConfig = new LTTngTestConfiguration()
            aggregatedConfig.TestBed <- baseConfig.TestBed
            aggregatedConfig.RunConfigurationList <-
                baseConfig.SDKVersionList
                |> List.map(
                    fun sdkVersion ->
                        let configuration = new LTTngTestRunConfiguration()

                        configuration.TestBed <- baseConfig.TestBed
                        configuration.TestResultFolder <- Path.Combine(configuration.TestBed, "TestResult")

                        configuration.DotNet.SDKVersion <- sdkVersion
                        configuration.DotNet.DotNetRoot <- Path.Combine(configuration.TestBed, $"dotnet-sdk-{sdkVersion}")

                        for de: DictionaryEntry in Environment.GetEnvironmentVariables() |> Seq.cast<DictionaryEntry> do
                            configuration.SystemInfo.EnvironmentVariables[de.Key.ToString()] <- de.Value.ToString()
                        if DotNet.CurrentRID.Contains("win")
                        then 
                            configuration.SystemInfo.EnvironmentVariables["Path"] <- 
                                Environment.GetEnvironmentVariable("Path") + $";{configuration.DotNet.DotNetRoot}"
                        else
                            configuration.SystemInfo.EnvironmentVariables["Path"] <- 
                                Environment.GetEnvironmentVariable("Path") + $":{configuration.DotNet.DotNetRoot}"
                        configuration.SystemInfo.EnvironmentVariables["DOTNET_ROOT"] <- configuration.DotNet.DotNetRoot

                        let targetAppsRoot = Path.Combine(configuration.TestBed, "TargetApps")
                        configuration.TargetApp.BuildConfig <- baseConfig.TargetApp.BuildConfig
                        configuration.TargetApp.GCPerfsim <- new DotNetApp.DotNetApp(configuration.SystemInfo.EnvironmentVariables,
                                                                                    "console",
                                                                                    Path.Combine(targetAppsRoot, $"gcperfsim-sdk{sdkVersion}-{DotNet.CurrentRID}"))
                        configuration.PerfCollectPath <- Path.Combine(configuration.TestBed, "perfcollect")
                        configuration)

            aggregatedConfig
             

File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Configuration/LTTngTestRunConfiguration.cs
using YamlDotNet.Serialization;

using DiagToolsValidationRunner.Core.Functionality;
using System.Collections;

namespace DiagToolsValidationRunner.Core.Configuration.LTTngTest
{
    public class TargetAppSetting : BaseTargetAppSetting
    {
        public DotNetApp? GCPerfsim;
    }

    public class LTTngTestRunConfiguration
    {
        public required BaseTestSetting Test;
        public required DotNetSDKSetting SDKSetting;
        public required TargetAppSetting AppSetting;
        public required BaseSystemInformation SysInfo;
    }

    public class LTTngTestConfiguration
    {
        public required BaseTestSetting Test;
        public required List<string> SDKVersionList;
        public required TargetAppSetting AppSetting;
        public List<LTTngTestRunConfiguration>? LTTngRunConfigurationList;
    }

    public static class LTTngTestConfigurationGenerator
    {
        private static IDeserializer _deserializer = (new DeserializerBuilder()).IgnoreUnmatchedProperties().Build();
        
        private static LTTngTestConfiguration ParseConfigFile(string configFile)
        {
            try
            {
                string serializedConfiguration = File.ReadAllText(configFile);
                LTTngTestConfiguration baseConfiguration =
                    _deserializer.Deserialize<LTTngTestConfiguration>(configFile);

                if (string.IsNullOrEmpty(baseConfiguration.Test.TestBed))
                {
                    throw new Exception(
                        $"{nameof(LTTngTestConfigurationGenerator)}: Please specify testbed");
                }

                else if (!(new List<string> { "Debug", "Release" }.Contains(baseConfiguration.AppSetting.BuildConfig)))
                {
                    throw new Exception(
                        $"{nameof(LTTngTestConfigurationGenerator)}: Please specify valid build config");
                }

                else if (baseConfiguration.SDKVersionList.Count == 0)
                {
                    throw new Exception(
                        $"{nameof(LTTngTestConfigurationGenerator)}: Please specify at least one SDK Version");
                }
                
                return baseConfiguration;
            }
            catch
            {
                throw;
            }
        }

        public static LTTngTestConfiguration GenerateConfiguration(string configFile)
        {
            LTTngTestConfiguration configuration = ParseConfigFile(configFile);
            string testResultFolder = Path.Combine(configuration.Test.TestBed, "TestResult");

            configuration.LTTngRunConfigurationList = new();
            foreach (string SDKVersion in configuration.SDKVersionList)
            {
                string testName = $"LTTng-SDK{SDKVersion}";
                string dotNetRoot = Path.Combine(configuration.Test.TestBed, $"DotNetSDK-{SDKVersion}");

                Dictionary<string, string> env = new();
                foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
                {
                    env[de!.Key!.ToString()!] = de!.Value!.ToString()!;
                }
                env["DOTNET_ROOT"] = dotNetRoot;

                string targetAppsRoot = Path.Combine(configuration.Test.TestBed, "TargetApps");
                string appRoot = Path.Combine(targetAppsRoot, $"gcperfsim-sdk{SDKVersion}-{DotNetInfrastructure.CurrentRID}");

                LTTngTestRunConfiguration runConfig = new()
                {
                    Test = new()
                    {
                        TestBed = configuration.Test.TestBed,
                        TestResultFolder = testResultFolder,
                        TestName = testName
                    },
                    SDKSetting = new()
                    {
                        Version = SDKVersion,
                        DotNetRoot = dotNetRoot,
                    },
                    SysInfo = new()
                    {
                        EnvironmentVariables = new(env)
                    },
                    AppSetting = new()
                    {
                        BuildConfig = configuration.AppSetting.BuildConfig,
                        GCPerfsim = new(env, "console", appRoot)
                    }
                };

                configuration.LTTngRunConfigurationList.Add(runConfig);
            }

            return configuration;
        }
    }
}
